### PR TITLE
Added initial yubikey support

### PR DESCRIPTION
**Pull request by Schwertspize**
*"Euda", Sights* (I just couldn't add a description :smile:)

this adds the possibility of loggin into rcon via a yubikey OTP (one time password). It is not considered as finished, it's a documentation of my progress.

1. What is "Bracestyle"?
2. It (for now) askes a system installed binary to verify the OTP, if this returns 0, then sets the rcon password to the OTP to avoid spoof protection to prevent the player from using rcon. as i discussed with @east in irc, a way better way is to add the OTP as "second" valid password and remove it after logout/leave. (This is no security hole, yubikey tokes are 45 characters long, thus i extended the rcon length, example: `cccccceegknghecfidvrjjuiljddhggekhvnbttnhjvn`)
3. explaining the config: the wanted OTP (45 char) is shrienked by the config reader to the first 12 chars, these 12 chars identify one yubikey, if you don't check it, everyone with A yubikey can login, not with THE one provided in the config (for convinience i ask for a full OTP because then the user dont have to count chars).

example OTP (you see my yubikey identification, the first 12 chars are always the same: `cccccceegkng`) (none of these are valid, because they are ONE TIME)
```
cccccceegkngujvvtdefukjfhlvgfgubvcevhgurrdnc
cccccceegkngivlcfiibbcbibrgunbhrtveehugvfgdj
cccccceegknggevjdbljkrefktjrnrgvdhvinnbifclv
cccccceegkngiehtvltglgvcihvufjguevkedfbnfftf
cccccceegkngkbcrbdggublkcghutjnhutrjcjifnjnr
cccccceegknggvejninkucdjtjdlculirjgtjbtnbdhl
cccccceegkngbhtetcnrbnkkkgjgiklitcvfviggnjnb
cccccceegkngjkrdrhicdurrergthlicrvdgbhtbkbed
cccccceegkngvlgruvnjgghrjnudvlhulcrtivfriclc
```

### CODE REVIEW

**Comment by H-M-H** [comment]
Would you be so kind and elaborate on what this is, how to use, how it works and why ddnet needs it ?

File: src/engine/server/server.cpp [hunk 1]
```diff
@@ -1154,6 +1155,26 @@ void CServer::ProcessClientPacket(CNetChunk *pPacket)
 
 			if((pPacket->m_Flags&NET_CHUNKFLAG_VITAL) != 0 && Unpacker.Error() == 0)
 			{
+
+				char yubikey_token[13];
+
+				memcpy( yubikey_token, &pPw[0], 12 );
+				yubikey_token[12] = '\0';
+
+				if(str_comp(yubikey_token, g_Config.m_SvYubikeyToken) == 0 && g_Config.m_SvUseYubikey == 1) {
```
Comment (H-M-H) [inline src/engine/server/server.cpp]: Bracestyle should not be changed.

File: src/engine/server/server.cpp [hunk 2]
```diff
@@ -1154,6 +1155,26 @@ void CServer::ProcessClientPacket(CNetChunk *pPacket)
...
+					int CLIENT_ID = 26493;
```
Comment (H-M-H) [inline src/engine/server/server.cpp]: I guess this is here as it states initial ?

File: src/engine/server/server.cpp [hunk 3]
```diff
@@ -1154,6 +1155,26 @@ void CServer::ProcessClientPacket(CNetChunk *pPacket)
 
 			if((pPacket->m_Flags&NET_CHUNKFLAG_VITAL) != 0 && Unpacker.Error() == 0)
 			{
+
+				char yubikey_token[13];
+
+				memcpy( yubikey_token, &pPw[0], 12 );
```
Comment (H-M-H) [inline src/engine/server/server.cpp]: mem_copy from base/system.h should be used. The same applies to things like strcpy, better use str_copy.

**Comment by def-** [comment]
No explanation for why we need this.

**Pull request closed without merging**